src/txdb.cpp

```diff
@@ -424,3 +425,129 @@ bool CCoinsViewDB::Upgrade() {
+}
+
+bool TxIndexDB::WriteTxns(const std::vector<std::pair<uint256, CDiskTxPos>>& v_pos)
+{
+    CDBBatch batch(*this);
+    for (auto tuple : v_pos) {
+        batch.Write(std::make_pair(DB_TXINDEX, tuple.first), tuple.second);
+    }
+    return WriteBatch(batch);
+}
+
+bool TxIndexDB::ReadBestBlockHash(uint256& hash) const {
+    if (Read(DB_BEST_BLOCK, hash)) {
+        return true;
+    }
+
+    // Read might have failed either because key does not exist or due to an error.
+    // If the former, return value should still be true.
+    if (!Exists(DB_BEST_BLOCK)) {
+        hash.SetNull();
+        return true;
+    }
+    return false;
+}
+
+bool TxIndexDB::WriteBestBlockHash(const uint256& hash) {
+    return Write(DB_BEST_BLOCK, hash);
+}
+
+bool TxIndexDB::MigrateData(CBlockTreeDB& block_tree_db, const uint256& block_hash)
+{
+    // The prior implementation of txindex was always in sync with block index
+    // and presence was indicated with a boolean DB flag.
+    bool f_migrate_index;
+    block_tree_db.ReadFlag("txindex", f_migrate_index);
+    if (!f_migrate_index) {
+        return true;
+    }
+
+    int64_t count = 0;
+    LogPrintf("Upgrading txindex database...\n");
+    LogPrintf("[0%%]...");
+    int report_done = 0;
+    size_t batch_size = 1 << 24;
+
+    CDBBatch batch_newdb(*this);
+    CDBBatch batch_olddb(block_tree_db);
+
+    std::pair<unsigned char, uint256> key;
+    std::pair<unsigned char, uint256> begin_key{DB_TXINDEX, uint256()};
+    std::pair<unsigned char, uint256> prev_key_newdb = begin_key;
+    std::pair<unsigned char, uint256> prev_key_olddb = begin_key;
+
+    std::unique_ptr<CDBIterator> pcursor(block_tree_db.NewIterator());
+    for (pcursor->Seek(begin_key); pcursor->Valid(); pcursor->Next()) {
+        boost::this_thread::interruption_point();
+        if (ShutdownRequested()) {
+            break;
+        }
+
+        if (!pcursor->GetKey(key)) {
+            break;
+        }
+        if (key.first != DB_TXINDEX) {
+            break;
+        }
+
+        // Log progress every 10%.
+        if (++count % 256 == 0) {
+            uint32_t high = 0x100 * *key.second.begin() + *(key.second.begin() + 1);
+            int percentage_done = (int)(high * 100.0 / 65536.0 + 0.5);
+            if (report_done < percentage_done/10) {
+                LogPrintf("[%d%%]...", percentage_done);
+                report_done = percentage_done/10;
+            }
+        }
+
+        CDiskTxPos value;
+        if (!pcursor->GetValue(value)) {
+            return error("%s: cannot parse txindex record", __func__);
+        }
+        batch_newdb.Write(key, value);
+        batch_olddb.Erase(key);
+
+        if (batch_newdb.SizeEstimate() > batch_size) {
+            WriteBatch(batch_newdb);
+            batch_newdb.Clear();
+            CompactRange(prev_key_newdb, key);
+            prev_key_newdb = key;
+        }
+        if (batch_olddb.SizeEstimate() > batch_size) {
```

Comment:
Separating these if statements loses coherency - if we crash we may have a bogus txindex and no way to detect it. Should likely instead do if (batch_newdb.SizeEstimate() + batch_olddb.SizeEstimate() > batch_size)........

Reply:
Good catch.